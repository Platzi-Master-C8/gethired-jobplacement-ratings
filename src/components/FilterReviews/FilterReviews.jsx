import React, { useState } from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import ArrowDownwardIcon from '@mui/icons-material/ArrowDownward';
import ArrowUpwardIcon from '@mui/icons-material/ArrowUpward';
import { Box, Button, InputLabel, FormControl, MenuItem, TextField, Typography, Select } from '@mui/material';

const FilterReviewsWrapper = styled.div`
    background-color: #ddd;
    border-radious: 1rem;
    margin-bottom: 1rem;
    padding: 1rem;
`;

const TextFieldS = styled(TextField)`
    background-color: #fff;
    border-radius: 4px;
    margin: 0 1rem;
    width: 50%;
    @media (max-width: 768px) {
        width: 40%;
    }
    @media (max-width: 480px) {
        width: 100%;
        margin: 1rem 0;
    }
`;

const FormControlS = styled(FormControl)`
    background-color: #fff;
    border-radius: 4px;
    width: 35%;
    @media (max-width: 768px) {
        width: 30%;
    }
    @media (max-width: 480px) {
        width: 100%;
    }
`;

const RowBox = styled(Box)`
    display: flex;
    align-items: center;
    flex-wrap: wrap;
    @media (max-width: 480px) {
        flex-direction: column;
    }
`;

const SortBox = styled(RowBox)`
    @media (max-width: 480px) {
        flex-direction: row;
        margin-top: 1rem;
    }
`;

const FilterReviews = ({ handleSearch, sortCriteria, toggleSortCriteria }) => {
    const [filterValue, setFilterValue] = useState('job_title');
    const [searchQuery, setSearchQuery] = useState('');

    const handleEnter = (e) => {
        if (e.key === 'Enter') handleSearch(searchQuery, filterValue);
    };

    return (
        <FilterReviewsWrapper>
            <Box
                sx={{
                    display: 'flex',
                    flexDirection: 'column',
                    borderBottom: '2px solid #FFF',
                }}
            >
                <Typography variant="subtitle1" sx={{ mb: 2 }}>
                    Search Reviews
                </Typography>
                <RowBox sx={{ mb: 2 }}>
                    <FormControlS>
                        <InputLabel id="filter-select-label">Search criteria</InputLabel>
                        <Select
                            labelId="filter-select-label"
                            id="filter-select"
                            value={filterValue}
                            label="Search criteria"
                            onChange={(e) => setFilterValue(e.target.value)}
                        >
                            <MenuItem value="job_title">Job title</MenuItem>
                            <MenuItem value="content_type">Review content</MenuItem>
                            <MenuItem value="job_location">Location</MenuItem>
                        </Select>
                    </FormControlS>
                    <TextFieldS
                        id="search-input"
                        label="Search Reviews"
                        variant="outlined"
                        value={searchQuery}
                        onChange={(e) => setSearchQuery(e.target.value)}
                        onKeyPress={handleEnter}
                    />
                    <Button variant="contained" onClick={() => handleSearch(searchQuery, filterValue)}>
                        Search
                    </Button>
                </RowBox>
            </Box>
            <RowBox sx={{ mt: 2 }}>
                <Typography variant="subtitle1" sx={{ mr: 2 }}>
                    Sort by:
                </Typography>
                <SortBox>
                    <Button
                        variant={sortCriteria.sortKey === 'utility_counter' ? 'contained' : 'outlined'}
                        endIcon={
                            sortCriteria.sortKey === 'utility_counter' ? (
                                sortCriteria.orientation === 'desc' ? (
                                    <ArrowUpwardIcon />
                                ) : (
                                    <ArrowDownwardIcon />
                                )
                            ) : null
                        }
                        onClick={() => toggleSortCriteria('utility_counter')}
                    >
                        Helpfulness
                    </Button>
                    <Button
                        variant={sortCriteria.sortKey === 'weighted_average_per_evaluation' ? 'contained' : 'outlined'}
                        endIcon={
                            sortCriteria.sortKey === 'weighted_average_per_evaluation' ? (
                                sortCriteria.orientation === 'desc' ? (
                                    <ArrowUpwardIcon />
                                ) : (
                                    <ArrowDownwardIcon />
                                )
                            ) : null
                        }
                        onClick={() => toggleSortCriteria('weighted_average_per_evaluation')}
                    >
                        Rating
                    </Button>
                    <Button
                        variant={sortCriteria.sortKey === 'created_at' ? 'contained' : 'outlined'}
                        endIcon={
                            sortCriteria.sortKey === 'created_at' ? (
                                sortCriteria.orientation === 'desc' ? (
                                    <ArrowUpwardIcon />
                                ) : (
                                    <ArrowDownwardIcon />
                                )
                            ) : null
                        }
                        onClick={() => toggleSortCriteria('created_at')}
                    >
                        Date
                    </Button>
                </SortBox>
            </RowBox>
        </FilterReviewsWrapper>
    );
};

FilterReviews.propTypes = {
    handleSearch: PropTypes.func.isRequired,
    sortCriteria: PropTypes.shape({
        sortKey: PropTypes.string.isRequired,
        orientation: PropTypes.string.isRequired,
    }).isRequired,
    toggleSortCriteria: PropTypes.func.isRequired,
};

export default FilterReviews;
