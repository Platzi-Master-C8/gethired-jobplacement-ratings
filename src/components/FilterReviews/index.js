import FilterReviews from './FilterReviews';

export {
    FilterReviews
}