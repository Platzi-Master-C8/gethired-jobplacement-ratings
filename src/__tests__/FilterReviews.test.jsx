import React from 'react';

import * as driver from '@testing-library/react';
import { FilterReviews } from 'Components/FilterReviews';

const defaultProps = {
    handleSearch: () => {},
    sortCriteria: { sortKey: 'created_at', orientation: 'asc' },
    toggleSortCriteria: () => {},
};

const renderComponent = (props = {}) => driver.render(<FilterReviews {...defaultProps} {...props} />);

describe('Component <FilterReviews/>', () => {
    test('renders FilterReviews', () => {
        const { asFragment } = renderComponent({});
        expect(asFragment()).toMatchSnapshot();
    });

    test('it allows write the search query', () => {
        const subject = renderComponent({});

        const searchInput = subject.container.querySelector('#search-input');
        driver.fireEvent.change(searchInput, { target: { value: '1234' } });

        expect(subject.getByDisplayValue('1234')).toBeInTheDocument();
    });

    test('it allows to search the writed text', () => {
        const handleSearch = jest.fn();
        const subject = renderComponent({ handleSearch });

        const searchInput = subject.container.querySelector('#search-input');
        const searchButton = subject.getByText('Search');
        driver.fireEvent.change(searchInput, { target: { value: '1234' } });
        driver.fireEvent.click(searchButton);

        expect(handleSearch).toBeCalled();
    });

    it('it searchs the user press the Enter key', () => {
        const handleSearch = jest.fn();
        const subject = renderComponent({ handleSearch });

        const searchInput = subject.container.querySelector('#search-input');

        driver.fireEvent.keyPress(searchInput, {
            key: 'Enter',
            code: 'Enter',
            charCode: 13,
        });

        expect(handleSearch).toBeCalled();
    });

    test('it allows to sort the list of reviews by Helpfulness', () => {
        const toggleSortCriteria = jest.fn();
        const sortCriteria = { sortKey: 'utility_counter', orientation: 'asc' };
        const subject = renderComponent({ toggleSortCriteria, sortCriteria });

        const sortByHelpfulnessButton = subject.container.querySelectorAll('button')[1];
        const arrowDownwardIcon = subject.getByTestId('ArrowDownwardIcon');

        expect(sortByHelpfulnessButton.querySelector('svg')).toEqual(arrowDownwardIcon);

        driver.fireEvent.click(sortByHelpfulnessButton);

        expect(toggleSortCriteria).toBeCalled();
    });

    test('once the list is sorted the helpfulness the button changes', () => {
        const sortCriteria = { sortKey: 'utility_counter', orientation: 'desc' };
        const subject = renderComponent({ sortCriteria });

        const sortByHelpfulnessButton = subject.container.querySelectorAll('button')[1];
        const arrowUpwardIcon = subject.getByTestId('ArrowUpwardIcon');

        expect(sortByHelpfulnessButton.querySelector('svg')).toEqual(arrowUpwardIcon);
    });

    test('it allows to sort the list of reviews by Rating', () => {
        const toggleSortCriteria = jest.fn();
        const sortCriteria = { sortKey: 'rating', orientation: 'asc' };
        const subject = renderComponent({ toggleSortCriteria, sortCriteria });

        const sortByRaitingButton = subject.container.querySelectorAll('button')[2];
        const arrowDownwardIcon = subject.getByTestId('ArrowDownwardIcon');

        expect(sortByRaitingButton.querySelector('svg')).toEqual(arrowDownwardIcon);

        driver.fireEvent.click(sortByRaitingButton);

        expect(toggleSortCriteria).toBeCalled();
    });

    test('once the list is sorted the rating the button changes', () => {
        const sortCriteria = { sortKey: 'rating', orientation: 'desc' };
        const subject = renderComponent({ sortCriteria });

        const sortByRaitingButton = subject.container.querySelectorAll('button')[2];
        const arrowUpwardIcon = subject.getByTestId('ArrowUpwardIcon');

        expect(sortByRaitingButton.querySelector('svg')).toEqual(arrowUpwardIcon);
    });

    test('it allows to sort the list of reviews by Date', () => {
        const toggleSortCriteria = jest.fn();
        const subject = renderComponent({ toggleSortCriteria });

        const sortByDateButton = subject.container.querySelectorAll('button')[3];
        const arrowDownwardIcon = subject.getByTestId('ArrowDownwardIcon');

        expect(sortByDateButton.querySelector('svg')).toEqual(arrowDownwardIcon);

        driver.fireEvent.click(sortByDateButton);

        expect(toggleSortCriteria).toBeCalled();
    });

    test('once the list is sorted the date the button changes', () => {
        const sortCriteria = { sortKey: 'created_at', orientation: 'desc' };
        const subject = renderComponent({ sortCriteria });

        const sortByDateButton = subject.container.querySelectorAll('button')[3];
        const arrowUpwardIcon = subject.getByTestId('ArrowUpwardIcon');

        expect(sortByDateButton.querySelector('svg')).toEqual(arrowUpwardIcon);
    });
});
