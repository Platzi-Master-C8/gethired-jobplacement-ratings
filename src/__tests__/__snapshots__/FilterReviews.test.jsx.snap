// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component <FilterReviews/> renders FilterReviews 1`] = `
<DocumentFragment>
  <div
    class="sc-bdvvtL ljDZpe"
  >
    <div
      class="MuiBox-root css-yugzle"
    >
      <h6
        class="MuiTypography-root MuiTypography-subtitle1 css-11j0rze-MuiTypography-root"
      >
        Search Reviews
      </h6>
      <div
        class="sc-hKwDye gptqDw MuiBox-root css-1qm1lh"
      >
        <div
          class="MuiFormControl-root sc-dkPtRN hhfCsF css-1nrlq1o-MuiFormControl-root"
        >
          <label
            class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary MuiFormLabel-filled css-1kty9di-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="true"
            id="filter-select-label"
          >
            Search criteria
          </label>
          <div
            class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <div
              aria-expanded="false"
              aria-haspopup="listbox"
              aria-labelledby="filter-select-label filter-select"
              class="MuiSelect-select MuiSelect-outlined MuiOutlinedInput-input MuiInputBase-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
              id="filter-select"
              role="button"
              tabindex="0"
            >
              Job title
            </div>
            <input
              aria-hidden="true"
              class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
              tabindex="-1"
              value="job_title"
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
              data-testid="ArrowDropDownIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-1z7n62"
              >
                <span>
                  Search criteria
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <div
          class="MuiFormControl-root MuiTextField-root sc-gsDKAQ ciUbap css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for="search-input"
            id="search-input-label"
          >
            Search Reviews
          </label>
          <div
            class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              id="search-input"
              type="text"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-173wfxe"
              >
                <span>
                  Search Reviews
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <button
          class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-sghohy-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          Search
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
    <div
      class="sc-hKwDye gptqDw MuiBox-root css-1yuhvjn"
    >
      <h6
        class="MuiTypography-root MuiTypography-subtitle1 css-3nv4t4-MuiTypography-root"
      >
        Sort by:
      </h6>
      <div
        class="sc-hKwDye sc-eCImPb gptqDw fuQCiZ MuiBox-root css-0"
      >
        <button
          class="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonBase-root  css-1rwt2y5-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          Helpfulness
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <button
          class="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonBase-root  css-1rwt2y5-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          Rating
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <button
          class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-sghohy-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          Date
          <span
            class="MuiButton-endIcon MuiButton-iconSizeMedium css-9tj150-MuiButton-endIcon"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="ArrowDownwardIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="m20 12-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
